File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章四阶幻方.cpp
//四阶幻方 2015年第六届决赛 lanqiaoOJ 689
#include <bits/stdc++.h>
using namespace std;
int vis[17];
int m[5][5];
int cnt = 0;
bool judge() {
    for (int i = 0; i < 4; i ++ ) { //四行四列的和
        if (m[i][0] + m[i][1] + m[i][2] + m[i][3] != 34) return 0;
        if (m[0][i] + m[1][i] + m[2][i] + m[3][i] != 34) return 0;
    }
    if (m[0][0] + m[1][1] + m[2][2] + m[3][3] != 34) return 0;//对角线
    if (m[0][3] + m[1][2] + m[2][1] + m[3][0] != 34) return 0;
    return 1;
}
void dfs(int n) {
    if (n == 16) {
        if (judge())  cnt ++;
        return;
    }
    if (n%4 == 0)       //简单剪枝：检查第一行的和是不是34
      if (m[n/4 - 1][0] + m[n/4 - 1][1] + m[n/4 - 1][2] + m[n/4 - 1][3] != 34)
         return;

    for (int i = 2; i <= 16; i ++ ) {
        if (vis[i] == 0) {
            m[n/4][n%4] = i;
            vis[i] = 1;
            dfs(n + 1);
            vis[i] = 0;
        }
    }
}
int main() {
     m[0][0] = 1;
     dfs(1);
     cout << cnt;
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章四阶幻方.py
#四阶幻方 2015年第六届决赛  lanqiaoOJ 689

def dfs(n):
    global cnt
    if n >= 4  and  m[0] + m[1]  + m[2]  + m[3]  != 34: return
    if n >= 7  and  m[0] + m[4]  + m[5]  + m[6]  != 34: return
    if n >= 10 and  m[1] + m[7]  + m[8]  + m[9]  != 34: return
    if n >= 11 and  m[3] + m[6]  + m[8]  + m[10] != 34: return
    if n >= 12 and  m[4] + m[7]  + m[10] + m[11] != 34: return
    if n >= 14 and  m[5] + m[8]  + m[12] + m[13] != 34: return
    if n >= 15 and  m[2] + m[10] + m[12] + m[14] != 34: return
    if n >= 16 and (m[6] + m[9]  + m[14] + m[15] != 34   \
                 or m[3] + m[11] + m[13] + m[15] != 34   \
                 or m[0] + m[7]  + m[12] + m[15] != 34):return
    if n == 16: cnt += 1
    for i in range(2, 17):  #2~16的排列
        if vis[i] == 0:            
            m[n]=i         
            vis[i] = 1
            dfs(n + 1)
            vis[i] = 0

cnt = 0 
m=[0]* 17
m[0]=1            #1被固定
vis = [0] * 17
vis[1] = 1
dfs(1)
print(cnt)
